lib/protocol/protocol.go

```diff
@@ -801,22 +809,42 @@ func (c *rawConnection) shouldCompressMessage(msg message) bool {
 	}
 }
 
+// Close is called when the connection is regularely closed and thus the Close
+// BEP message is sent before terminating the actual connection. The error
+// argument specifies the reason for closing the connection.
+func (c *rawConnection) Close(err error) {
+	c.once.Do(func() {
+		done := make(chan struct{})
+		c.sendClose <- asyncMessage{&Close{err.Error()}, done}
+		<-done
+
+		go c.commonClose(err)
+	})
+}
+
+// close is called if there is an unexpected error during normal operation.
 func (c *rawConnection) close(err error) {
```

Comment:
I'm not super fond of the having both `close()` and `Close()`, the potential for confusion is enormous... I'm slightly confused even now. 😕 <- look, confused emoji

Reply:
Is `internalClose` suitable? It's a bit cryptic, but so is `errorClose` when `Close` takes an error as an argument as well and I couldn't think of anything better.